src/lavinmq/exchange/exchange.cr

```diff
@@ -167,18 +144,83 @@ module LavinMQ
 
     abstract def type : String
-    abstract def bind(destination : Queue, routing_key : String, headers : AMQP::Table?)
-    abstract def unbind(destination : Queue, routing_key : String, headers : AMQP::Table?)
-    abstract def bind(destination : Exchange, routing_key : String, headers : AMQP::Table?)
-    abstract def unbind(destination : Exchange, routing_key : String, headers : AMQP::Table?)
-    abstract def do_queue_matches(routing_key : String, headers : AMQP::Table?, & : Queue -> _)
-    abstract def do_exchange_matches(routing_key : String, headers : AMQP::Table?, & : Exchange -> _)
-
-    def queue_matches(routing_key : String, headers = nil, &blk : Queue -> _)
+    abstract def bind(destination : Destination, routing_key : String, headers : AMQP::Table?)
+    abstract def unbind(destination : Destination, routing_key : String, headers : AMQP::Table?)
+    abstract def bindings_details : Iterator(BindingDetails)
+
+    def publish(msg : Message, immediate : Bool,
+                queues : Set(Queue) = Set(Queue).new,
+                exchanges : Set(Exchange) = Set(Exchange).new) : Int32
+      @publish_in_count += 1
+      headers = msg.properties.headers
       if should_delay_message?(headers)
-        @delayed_queue.try { |q| yield q }
-      else
-        do_queue_matches(routing_key, headers, &blk)
+        if q = @delayed_queue
+          q.publish(msg)
+          return 1
+        else
+          return 0
+        end
+      end
+      find_queues(msg.routing_key, headers, queues, exchanges)
+      if queues.empty?
+        @unroutable_count += 1
+        return 0
+      end
+      return 0 if immediate && !queues.any? &.immediate_delivery?
+
+      count = 0
+      queues.each do |queue|
+        if queue.publish(msg)
+          @publish_out_count += 1
+          count += 1
+          msg.body_io.seek(-msg.bodysize.to_i64, IO::Seek::Current) # rewind
+        end
+      end
+      count
+    end
+
+    def find_queues(routing_key : String, headers : AMQP::Table?,
+                    queues : Set(Queue),
+                    exchanges : Set(Exchange) = Set(Exchange).new) : Nil
+      return unless exchanges.add? self
+      bindings(routing_key, headers).each do |d|
+        case d
+        when Queue
+          queues.add(d) unless d.closed?
+        when Exchange
+          d.find_queues(routing_key, headers, queues, exchanges)
+        end
+      end
+
+      if hdrs = headers
+        find_cc_queues(hdrs, "CC", queues)
+        find_cc_queues(hdrs, "BCC", queues)
+        hdrs.delete "BCC"
+      end
+
+      if queues.empty? && alternate_exchange
+        @vhost.exchanges[alternate_exchange]?.try do |ae|
+          ae.find_queues(routing_key, headers, queues, exchanges)
+        end
+      end
+    end
+
+    private def find_cc_queues(headers, key, queues)
+      return unless cc = headers[key]?
+      cc = cc.as?(Array(AMQP::Field))
+
+      err = Error::PreconditionFailed.new("#{key} header not a string array")
```

Comment:
This means that we are always allocating, even when never raising

Reply:
Extract it to a method (2 extra rows), or duplicate it (1 extra row). Allocations in the hot path is bad.

Reply:
Will just copy the line